File: backend/app/routers/cashflow.py
from fastapi import APIRouter
from pydantic import BaseModel, Field
from typing import List, Optional, Literal


router = APIRouter()


class LineItem(BaseModel):
    name: str = Field(..., min_length=1, max_length=64)
    amount: float = Field(..., ge=0)  # non-negative cash amounts


class CashflowRequest(BaseModel):
    """
    Precise monthly cash flow inputs. Amounts are monthly and in the same
    currency. Taxes are approximated as a cash outflow on positive
    operating profit.
    """

    starting_cash: float = Field(0, ge=0)
    revenue: List[LineItem] = Field(..., min_length=1)
    other_inflows: List[LineItem] = Field(default_factory=list)

    # Cash outflows (monthly)
    cogs: List[LineItem] = Field(default_factory=list)
    opex_fixed: List[LineItem] = Field(default_factory=list)
    opex_variable: List[LineItem] = Field(default_factory=list)
    payroll: List[LineItem] = Field(default_factory=list)
    loan_interest: List[LineItem] = Field(default_factory=list)
    loan_principal: List[LineItem] = Field(default_factory=list)
    capex: List[LineItem] = Field(default_factory=list)
    other_outflows: List[LineItem] = Field(default_factory=list)

    # effective monthly tax rate on positive operating profit
    tax_rate: float = Field(0.0, ge=0, le=1)
    round_to: int = Field(2, ge=0, le=6)


class CashflowTotals(BaseModel):
    total_inflows: float
    total_outflows: float
    taxes_cash: float
    operating_cash_flow: float
    free_cash_flow: float
    net_cash_flow: float
    ending_cash: float
    burn_rate: float
    runway_months: Optional[float]


class CashflowBreakdown(BaseModel):
    inflows: List[LineItem]
    outflows: List[LineItem]


class CashflowAnalysis(BaseModel):
    summary: str
    risk_level: Literal["low", "medium", "high"]
    recommendations: List[str]


class CashflowResponse(BaseModel):
    totals: CashflowTotals
    breakdown: CashflowBreakdown
    analysis: CashflowAnalysis


def _sum(items: List[LineItem]) -> float:
    return float(sum(i.amount for i in items))


def _round(v: float, nd: int) -> float:
    return round(float(v), nd)


@router.post("/calculate", response_model=CashflowResponse)
def calculate_cashflow(payload: CashflowRequest) -> CashflowResponse:
    # Inflows
    rev = _sum(payload.revenue)
    other_in = _sum(payload.other_inflows)
    total_inflows = rev + other_in

    # Operating outflows (cash)
    cogs = _sum(payload.cogs)
    opex_fix = _sum(payload.opex_fixed)
    opex_var = _sum(payload.opex_variable)
    payroll = _sum(payload.payroll)
    operating_cash_out = cogs + opex_fix + opex_var + payroll

    # Financing/Investing and other outflows
    interest = _sum(payload.loan_interest)
    principal = _sum(payload.loan_principal)
    capex = _sum(payload.capex)
    other_out = _sum(payload.other_outflows)

    # Approximate taxes only when there is positive operating profit
    operating_profit_approx = rev - operating_cash_out - interest
    taxes_cash = 0.0
    if operating_profit_approx > 0 and payload.tax_rate > 0:
        taxes_cash = operating_profit_approx * float(payload.tax_rate)

    # Totals
    total_outflows = (
        operating_cash_out
        + interest
        + principal
        + capex
        + other_out
        + taxes_cash
    )
    net_cash_flow = total_inflows - total_outflows
    ending_cash = payload.starting_cash + net_cash_flow

    operating_cash_flow = rev - operating_cash_out - taxes_cash
    free_cash_flow = operating_cash_flow - capex

    burn_rate = -net_cash_flow if net_cash_flow < 0 else 0.0
    runway_months = None
    if burn_rate > 0:
        runway_months = (
            payload.starting_cash / burn_rate if burn_rate else None
        )

    # Risk level heuristic
    risk_level: Literal["low", "medium", "high"]
    if net_cash_flow >= 0 and (runway_months is None or runway_months >= 6):
        risk_level = "low"
    elif net_cash_flow >= 0 or (
        runway_months is not None and runway_months >= 3
    ):
        risk_level = "medium"
    else:
        risk_level = "high"

    # Analysis & recommendations
    rnd = payload.round_to
    if net_cash_flow >= 0:
        summary = (
            f"Flujo de caja positivo de {_round(net_cash_flow, rnd)}. "
            f"Saldo final {_round(ending_cash, rnd)}. "
            f"Flujo operativo {_round(operating_cash_flow, rnd)}; "
            f"FCF {_round(free_cash_flow, rnd)}."
        )
        recommendations = [
            (
                "Considera destinar 10-20% del flujo positivo a un fondo de "
                "reserva (runway)."
            ),
            "Evalúa reducir deuda con mayor tasa utilizando el excedente.",
            (
                "Reinvierte en canales con ROI probado (p. ej., campañas con "
                "CAC < LTV)."
            ),
        ]
    else:
        summary = (
            f"Flujo de caja negativo de {_round(-net_cash_flow, rnd)} (burn). "
            f"Runway estimado: "
            f"{_round(runway_months, rnd) if runway_months else 0} meses."
        )
        # Identify top outflow buckets
        buckets = [
            ("COGS", cogs),
            ("OPEX fijo", opex_fix),
            ("OPEX variable", opex_var),
            ("Nómina", payroll),
            ("Intereses", interest),
            ("Principal deuda", principal),
            ("CapEx", capex),
            ("Otros", other_out),
        ]
        buckets.sort(key=lambda x: x[1], reverse=True)
        top = ", ".join(
            [f"{n}: {_round(v, rnd)}" for n, v in buckets[:3] if v > 0]
        )
        rec_delta_rev = max(0.0, total_outflows - total_inflows)
        recommendations = [
            (
                f"Prioriza recortes en: {top}"
                if top
                else (
                    "Revisa cada partida de gasto para identificar recortes."
                )
            ),
            "Negocia pagos con proveedores o alarga plazos para aliviar caja.",
            (
                "Objetivo de aumento de ingresos: "
                f"{_round(rec_delta_rev, rnd)} para alcanzar punto de "
                "equilibrio de caja."
            ),
        ]

    totals = CashflowTotals(
        total_inflows=_round(total_inflows, rnd),
        total_outflows=_round(total_outflows, rnd),
        taxes_cash=_round(taxes_cash, rnd),
        operating_cash_flow=_round(operating_cash_flow, rnd),
        free_cash_flow=_round(free_cash_flow, rnd),
        net_cash_flow=_round(net_cash_flow, rnd),
        ending_cash=_round(ending_cash, rnd),
        burn_rate=_round(burn_rate, rnd),
        runway_months=(
            _round(runway_months, rnd) if runway_months is not None else None
        ),
    )

    breakdown = CashflowBreakdown(
        inflows=[
            LineItem(name="Ingresos", amount=_round(rev, rnd)),
            LineItem(name="Otros ingresos", amount=_round(other_in, rnd)),
        ],
        outflows=[
            LineItem(name="COGS", amount=_round(cogs, rnd)),
            LineItem(name="OPEX fijo", amount=_round(opex_fix, rnd)),
            LineItem(name="OPEX variable", amount=_round(opex_var, rnd)),
            LineItem(name="Nómina", amount=_round(payroll, rnd)),
            LineItem(name="Impuestos (cash)", amount=_round(taxes_cash, rnd)),
            LineItem(name="Intereses", amount=_round(interest, rnd)),
            LineItem(name="Principal deuda", amount=_round(principal, rnd)),
            LineItem(name="CapEx", amount=_round(capex, rnd)),
            LineItem(name="Otros egresos", amount=_round(other_out, rnd)),
        ],
    )

    analysis = CashflowAnalysis(
        summary=summary,
        risk_level=risk_level,
        recommendations=recommendations,
    )

    return CashflowResponse(
        totals=totals, breakdown=breakdown, analysis=analysis
    )

File: src/lib/calculations/cashflow.ts
// Librería de cálculos de flujo de caja
export interface LineItem {
  name: string;
  amount: number;
}

export interface CashflowInput {
  startingCash: number;
  revenue: LineItem[];
  otherInflows: LineItem[];
  cogs: LineItem[];
  opexFixed: LineItem[];
  opexVariable: LineItem[];
  payroll: LineItem[];
  loanInterest: LineItem[];
  loanPrincipal: LineItem[];
  capex: LineItem[];
  otherOutflows: LineItem[];
  taxRate: number; // 0-1
  roundTo?: number;
}

export interface CashflowTotals {
  totalInflows: number;
  totalOutflows: number;
  taxesCash: number;
  operatingCashFlow: number;
  freeCashFlow: number;
  netCashFlow: number;
  endingCash: number;
  burnRate: number;
  runwayMonths: number | null;
}

export interface CashflowBreakdown {
  inflows: LineItem[];
  outflows: LineItem[];
}

export interface CashflowAnalysis {
  summary: string;
  riskLevel: 'low' | 'medium' | 'high';
  recommendations: string[];
}

export interface CashflowResult {
  totals: CashflowTotals;
  breakdown: CashflowBreakdown;
  analysis: CashflowAnalysis;
}

export class CashflowCalculator {
  /**
   * Calcula el flujo de caja mensual completo
   */
  static calculate(input: CashflowInput): CashflowResult {
    const roundTo = input.roundTo || 2;
    
    // Calcular totales de ingresos
    const revenue = this.sumItems(input.revenue);
    const otherInflows = this.sumItems(input.otherInflows);
    const totalInflows = revenue + otherInflows;
    
    // Calcular egresos operativos
    const cogs = this.sumItems(input.cogs);
    const opexFixed = this.sumItems(input.opexFixed);
    const opexVariable = this.sumItems(input.opexVariable);
    const payroll = this.sumItems(input.payroll);
    const operatingCashOut = cogs + opexFixed + opexVariable + payroll;
    
    // Calcular egresos financieros y de inversión
    const loanInterest = this.sumItems(input.loanInterest);
    const loanPrincipal = this.sumItems(input.loanPrincipal);
    const capex = this.sumItems(input.capex);
    const otherOutflows = this.sumItems(input.otherOutflows);
    
    // Calcular impuestos (solo sobre utilidad operativa positiva)
    const operatingProfitApprox = revenue - operatingCashOut - loanInterest;
    const taxesCash = operatingProfitApprox > 0 && input.taxRate > 0 
      ? operatingProfitApprox * input.taxRate 
      : 0;
    
    // Totales finales
    const totalOutflows = operatingCashOut + loanInterest + loanPrincipal + capex + otherOutflows + taxesCash;
    const netCashFlow = totalInflows - totalOutflows;
    const endingCash = input.startingCash + netCashFlow;
    
    const operatingCashFlow = revenue - operatingCashOut - taxesCash;
    const freeCashFlow = operatingCashFlow - capex;
    
    // Burn rate y runway
    const burnRate = netCashFlow < 0 ? -netCashFlow : 0;
    const runwayMonths = burnRate > 0 ? input.startingCash / burnRate : null;
    
    // Determinar nivel de riesgo
    let riskLevel: 'low' | 'medium' | 'high';
    if (netCashFlow >= 0 && (!runwayMonths || runwayMonths >= 6)) {
      riskLevel = 'low';
    } else if (netCashFlow >= 0 || (runwayMonths && runwayMonths >= 3)) {
      riskLevel = 'medium';
    } else {
      riskLevel = 'high';
    }
    
    // Generar análisis y recomendaciones
    const analysis = this.generateAnalysis(
      netCashFlow, endingCash, operatingCashFlow, freeCashFlow, 
      runwayMonths, riskLevel, roundTo, totalInflows, totalOutflows,
      { cogs, opexFixed, opexVariable, payroll, loanInterest, loanPrincipal, capex, otherOutflows }
    );
    
    // Crear totales redondeados
    const totals: CashflowTotals = {
      totalInflows: this.round(totalInflows, roundTo),
      totalOutflows: this.round(totalOutflows, roundTo),
      taxesCash: this.round(taxesCash, roundTo),
      operatingCashFlow: this.round(operatingCashFlow, roundTo),
      freeCashFlow: this.round(freeCashFlow, roundTo),
      netCashFlow: this.round(netCashFlow, roundTo),
      endingCash: this.round(endingCash, roundTo),
      burnRate: this.round(burnRate, roundTo),
      runwayMonths: runwayMonths ? this.round(runwayMonths, roundTo) : null
    };
    
    // Crear breakdown
    const breakdown: CashflowBreakdown = {
      inflows: [
        { name: "Ingresos", amount: this.round(revenue, roundTo) },
        { name: "Otros ingresos", amount: this.round(otherInflows, roundTo) }
      ],
      outflows: [
        { name: "COGS", amount: this.round(cogs, roundTo) },
        { name: "OPEX fijo", amount: this.round(opexFixed, roundTo) },
        { name: "OPEX variable", amount: this.round(opexVariable, roundTo) },
        { name: "Nómina", amount: this.round(payroll, roundTo) },
        { name: "Impuestos (cash)", amount: this.round(taxesCash, roundTo) },
        { name: "Intereses", amount: this.round(loanInterest, roundTo) },
        { name: "Principal deuda", amount: this.round(loanPrincipal, roundTo) },
        { name: "CapEx", amount: this.round(capex, roundTo) },
        { name: "Otros egresos", amount: this.round(otherOutflows, roundTo) }
      ]
    };
    
    return {
      totals,
      breakdown,
      analysis
    };
  }
  
  /**
   * Suma los elementos de una lista
   */
  private static sumItems(items: LineItem[]): number {
    return items.reduce((sum, item) => sum + (item.amount || 0), 0);
  }
  
  /**
   * Redondea un número
   */
  private static round(value: number, decimals: number): number {
    return Math.round(value * Math.pow(10, decimals)) / Math.pow(10, decimals);
  }
  
  /**
   * Genera análisis y recomendaciones
   */
  private static generateAnalysis(
    netCashFlow: number, endingCash: number, operatingCashFlow: number, 
    freeCashFlow: number, runwayMonths: number | null, riskLevel: string,
    roundTo: number, totalInflows: number, totalOutflows: number,
    buckets: {
      cogs: number;
      opexFixed: number;
      opexVariable: number;
      payroll: number;
      loanInterest: number;
      loanPrincipal: number;
      capex: number;
      otherOutflows: number;
    }
  ): CashflowAnalysis {
    
    let summary: string;
    let recommendations: string[];
    
    if (netCashFlow >= 0) {
      summary = `Flujo de caja positivo de ${this.round(netCashFlow, roundTo)}. ` +
               `Saldo final ${this.round(endingCash, roundTo)}. ` +
               `Flujo operativo ${this.round(operatingCashFlow, roundTo)}; ` +
               `FCF ${this.round(freeCashFlow, roundTo)}.`;
      
      recommendations = [
        "Considera destinar 10-20% del flujo positivo a un fondo de reserva (runway).",
        "Evalúa reducir deuda con mayor tasa utilizando el excedente.",
        "Reinvierte en canales con ROI probado (p. ej., campañas con CAC < LTV)."
      ];
    } else {
      summary = `Flujo de caja negativo de ${this.round(-netCashFlow, roundTo)} (burn). ` +
               `Runway estimado: ${runwayMonths ? this.round(runwayMonths, roundTo) : 0} meses.`;
      
      // Identificar los mayores egresos
      const expenseBuckets = [
        ["COGS", buckets.cogs],
        ["OPEX fijo", buckets.opexFixed],
        ["OPEX variable", buckets.opexVariable],
        ["Nómina", buckets.payroll],
        ["Intereses", buckets.loanInterest],
        ["Principal deuda", buckets.loanPrincipal],
        ["CapEx", buckets.capex],
        ["Otros", buckets.otherOutflows]
      ];
      
      expenseBuckets.sort((a, b) => (b[1] as number) - (a[1] as number));
      const topExpenses = expenseBuckets
        .filter(([, amount]) => (amount as number) > 0)
        .slice(0, 3)
        .map(([name, amount]) => `${name}: ${this.round(amount as number, roundTo)}`)
        .join(", ");
      
      const recommendedRevenueIncrease = Math.max(0, totalOutflows - totalInflows);
      
      recommendations = [
        topExpenses 
          ? `Prioriza recortes en: ${topExpenses}`
          : "Revisa cada partida de gasto para identificar recortes.",
        "Negocia pagos con proveedores o alarga plazos para aliviar caja.",
        `Objetivo de aumento de ingresos: ${this.round(recommendedRevenueIncrease, roundTo)} para alcanzar punto de equilibrio de caja.`
      ];
    }
    
    return {
      summary,
      riskLevel: riskLevel as 'low' | 'medium' | 'high',
      recommendations
    };
  }
}
